kernelflinger: efiwrapper: find rsdp in ebda area

zhangckid opened the pull request:
Per ACPI spec (5.2.5.1 Finding the RSDP on IA-PC Systems), OSPM could find RSDP by searching physical memory ranges on 16-byte boundaries at "The first 1 KB of
 the Extended BIOS Data Area (EBDA). For EISA or MCA systems, the EBDA can
be found in the two-byte location 40:0Eh on the BIOS data area."

Although EBDA is deprecated on modern UEFI platform, but it could exist on some virtualization platfroms. Since ACPI RSDP info is not supported by multiboot protocol so it might get lost by searching 0xe0000~0x100000 area only.

AjanZhong commented on drivers/acpi/acpi.c:
```diff
@@ -106,6 +107,11 @@ static struct RSDP_TABLE *lookup_for_rdsp(char *from)
 		if (!memcmp(p, RSDP_MAGIC, sizeof(RSDP_MAGIC)))
 			return (struct RSDP_TABLE *)p;
 
+	from = (char *)((*ebda_addr) << 4);
```
Just curious why do we need to left shift 4 to content resides in *ebda_addr, it looks like segment in real mode. Since "the EBDA can be found in the two-byte location 40:0Eh on the BIOS data area."

jsun26intel replied:
This is legacy way to scan RSDP with real mode. real mode 40:0Eh means 0x40e in physical address.
Please refer kernel implementation at drivers/acpi/acpica/tbxfroot.c:
acpi_status ACPI_INIT_FUNCTION
acpi_find_root_pointer(acpi_physical_address *table_address)
{
...
}

AjanZhong replied:
LGTM

AjanZhong replied:
As discussed afternoon, please add comment to indicate the reason why we need to left shift 4 bit of value which is read from EBDA.

jsun26intel replied:
The 0x40e stores EBDA segment address actually. I haven't found formal spec talk about this issue, but it is mentioned everywhere if you use google.

Here is wiki from https://wiki.osdev.org/Memory_Map_(x86)
0x040E (word) | EBDA base address >> 4 (usually!)


@zhangckid  please change "ebda_addr" to "ebda_seg_addr" and add a comment that:
/* 0x40e stores EBDA segment address, need left shift 4 bits to get EBDA physical address. */

Aryan12138 committed:
efiwrapper: find rsdp in ebda area

Per ACPI spec (5.2.5.1 Finding the RSDP on IA-PC Systems), OSPM could find RSDP
by searching physical memory ranges on 16-byte boundaries at "The first 1 KB of
 the Extended BIOS Data Area (EBDA). For EISA or MCA systems, the EBDA can
be found in the two-byte location 40:0Eh on the BIOS data area."

Although EBDA is deprecated on modern UEFI platform, but it could exist on some
virtualization platfroms. Since ACPI RSDP info is not supported by multiboot
protocol so it might get lost by searching 0xe0000~0x100000 area only.

Tracked-On: OAM-111350

Signed-off-by: Victor Sun <victor.sun@intel.com>
Signed-off-by: Zhang Chen <chen.zhang@intel.com>

zhangckid commented:
Update the patch as comments, please review again. Thanks.

AjanZhong reviewed (approved):
LGTM

GangSecurity reviewed (commented):
No tracked on message. 

Tracked-On: OAM-111036
